core/src/main/scala/spire/math/Algebraic.scala

```diff
@@ -66,86 +167,1383 @@ extends ScalaNumber with ScalaNumericConversions
 
-  implicit def apply(n: Int): Algebraic = Expr(n)
-  implicit def apply(n: Long): Algebraic = Expr(n)
-  implicit def apply(n: BigInt): Algebraic = Expr(n)
-  implicit def apply(n: Rational): Algebraic = Expr(n)
-  implicit def apply(n: Double): Algebraic = Expr(n)
-  implicit def apply(n: BigDecimal): Algebraic = Expr(n)
+  /** Returns an Algebraic expression equal to 0. */
+  val Zero: Algebraic = new Algebraic(Expr.ConstantLong(0))
+
+  /** Returns an Algebraic expression equal to 1. */
+  val One: Algebraic = new Algebraic(Expr.ConstantLong(1))
+
+  /** Returns an Algebraic expression equivalent to `n`. */
+  implicit def apply(n: Int): Algebraic =
+    new Algebraic(Expr.ConstantLong(n))
 
-  implicit object RealCoexpr extends Coexpr[Algebraic] {
-    def expr(r: Algebraic): Expr[Algebraic] = r.expr
-    def coexpr(e: Expr[Algebraic]): Algebraic = new Algebraic(e)
+  /** Returns an Algebraic expression equivalent to `n`. */
+  def apply(n: Long): Algebraic =
+    new Algebraic(Expr.ConstantLong(n))
+
+  /**
+   * Returns an Algebraic expression equivalent to `n`, if `n` is finite. If
+   * `n` is either infinite or `NaN`, then an `IllegalArgumentException` is
+   * thrown.
+   */
+  def apply(n: Float): Algebraic =
+    Algebraic(n.toDouble)
+
+  /**
+   * Returns an Algebraic expression equivalent to `n`, if `n` is finite. If
+   * `n` is either infinite or `NaN`, then an `IllegalArgumentException` is
+   * thrown.
+   */
+  implicit def apply(n: Double): Algebraic =
+    if (java.lang.Double.isInfinite(n)) {
+      throw new IllegalArgumentException("cannot construct inifinite Algebraic")
+    } else if (java.lang.Double.isNaN(n)) {
+      throw new IllegalArgumentException("cannot construct Algebraic from NaN")
+    } else {
+      new Algebraic(Expr.ConstantDouble(n))
+    }
+
+  /** Returns an Algebraic expression equivalent to `n`. */
+  def apply(n: BigInt): Algebraic =
+    new Algebraic(Expr.ConstantBigDecimal(BigDecimal(n)))
+
+  /** Returns an Algebraic expression equivalent to `n`. */
+  def apply(n: BigDecimal): Algebraic =
+    new Algebraic(Expr.ConstantBigDecimal(n))
+
+  /** Returns an Algebraic expression equivalent to `n`. */
+  def apply(n: Rational): Algebraic =
+    new Algebraic(Expr.ConstantRational(n))
+
+  /**
+   * Returns an Algebraic expression whose value is equivalent to the `i`-th
+   * real root of the [[Polynomial]] `poly`. If `i` is negative or does not an
+   * index a real root (eg the value is greater than or equal to the number of
+   * real roots) then an `ArithmeticException` is thrown. Roots are indexed
+   * starting at 0.  So if there are 3 roots, then they are indexed as 0, 1,
+   * and 2.
+   *
+   * @param poly the polynomial containing at least i real roots
+   * @param i    the index (0-based) of the root
+   * @return an algebraic whose value is the i-th root of the polynomial
+   */
+  def root(poly: Polynomial[Rational], i: Int): Algebraic = {
```

Comment:
Should add some tests to verify that calling this with an invalid index (< 0 or >= # of roots) throws an exception.